### PR TITLE
Nist registrar server

### PR DESCRIPTION
**Pull request by AshleySetter**
Added Regisrar server which runs on the registrar alongside a VC Rest API instance. Receives VCs through it's Rest API endpoint, verifies them using the VC Rest API instance and acts on verified VC's information to update the sqlite database, if entities referred to in the claim exist.

Implemented only manufacturer_trust for now.

### CODE REVIEW

File: packages/nist_registrar_server/README.md [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# Nist Registrar Server
+
+This is the REST API server which should run on the registrar and accept VCs containing claims, it verifies the VC with the VC Rest API and if the VC successfully verifies then the information claimed in the VCs is acted upon to update the information in sqlite database to reflect the claim.
+
+In order for this to work you must first following the instructions to set up the @nqminds/nist_vc_rest_server and get it running on the same machine as this server. In order for the server to work correctly the following fields in the config file must be pointed at the appropriate targets:
```
Comment (anthonymccaigue) [inline packages/nist_registrar_server/README.md:5]: Hi Ash,
Theres no instructions to set this up?

**Review by anthonymccaigue — COMMENTED** [review]
I wasn't able to test the code works as intended so just done a code review as best as possible

File: packages/nist_registrar_server/src/server.js [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+              } else {
+                console.log('Schema executed successfully');
+              }
+            });
+            // List of usernames
+            const usernames = ['Ash', 'John', 'Ant', 'James', 'Ionut', 'Nick', 'Anita', 'Hannah', 'Suzanne', 'Alois', 'Alex', 'Toby'];
+  
+            // Insert data into user table with UUID for each user
+            for (const username of usernames) {
+              db.run("INSERT INTO user (id, username, created_at) VALUES (?, ?, ?)", [uuidv4(), username, new Date().toISOString()]);
```
[the remark this answers is not in the document]
Comment (AshleySetter) [reply packages/nist_registrar_server/src/server.js:124]: VCs have been designed so you can use name or id, so id's don't need to be persisted

File: packages/nist_registrar_server/src/server.js [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+  
+            // Insert a single entry into gives_purchase_rights
+            const ashRow = db.get("SELECT id FROM user WHERE username = 'Ash'");
+            const ashsId = ashRow.id;
+  
+            const nickRowForRights = db.get("SELECT id FROM user WHERE username = 'Nick'");
+            const nicksIdForRights = nickRowForRights.id;
+  
+            db.run("INSERT INTO gives_purchase_rights (recipient_id, authoriser_id, created_at) VALUES (?, ?, ?)", [ashsId, nicksIdForRights, new Date().toISOString()]);
+          });
```
[the remark this answers is not in the document]
Comment (AshleySetter) [reply packages/nist_registrar_server/src/server.js:149]: yeah, just so we have some users in the database who can trust devices / manufacturers

File: packages/nist_registrar_server/src/server.js [hunk 4]
```diff
@@ -0,0 +1,270 @@
...
+       new URL("../config.json", import.meta.url),
+       {encoding: "utf8"}),
+     ));
+   }
+
+   if (config.sqliteDBPath) {
+    const sqliteDBPath = fileURLToPath(new URL(config.sqliteDBPath, import.meta.url))
+    try {
+      // Open a database connection
+      db = new sqlite3.Database(sqliteDBPath, sqlite3.OPEN_READWRITE, (err) => {
```
[the remark this answers is not in the document]
Comment (AshleySetter) [reply packages/nist_registrar_server/src/server.js:106]: The rust scripts which implement the policy use rusqlite, they're not set up to run off the volt.

File: packages/nist_registrar_server/package.json [hunk 5]
```diff
@@ -0,0 +1,29 @@
+{
+  "name": "nist_registrar_server",
```
Comment (anthonymccaigue) [inline packages/nist_registrar_server/package.json:2]:
```suggestion
  "name": "@nqminds/nist_registrar_server",
```

File: packages/nist_registrar_server/src/handle_device_type_binding.js [hunk 6]
```diff
@@ -0,0 +1,30 @@
+export default async function handle_device_type_binding(claimData, dbGet, dbRun) {
+  const {device, deviceType, issuanceDate} = claimData;device
```
Comment (anthonymccaigue) [inline packages/nist_registrar_server/src/handle_device_type_binding.js:2]:
```suggestion
  const {device, deviceType, issuanceDate} = claimData;
```

File: packages/nist_registrar_server/src/handle_device_type_vulnerable.js [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+  }
+  const demoVulnerability = await dbGet("SELECT id from vulnerability where id = ?", [demoVulnerabilityId]);
+  if (!demoVulnerability) {
+    await dbRun("INSERT INTO vulnerability (id, name, severity, url, created_at) VALUES (?, ?, ?, ?, ?)", [demoVulnerabilityId, 'Security Flaw',	'Critical',	'https://example.com', '2024-01-16 14:13:55.849160']);
+  }
+  const has_demo_vulnerability = await dbGet("SELECT * from has_vulnerability WHERE device_type_id = ? AND vulnerability_id = ?", [deviceTypeId, demoVulnerabilityId]);
+  console.log("has_demo_vuln: ", has_demo_vulnerability);
+  if (vulnerable) {
+    if (has_demo_vulnerability) {
+      return `device type ${deviceType} is already vulnerable`;
```
Comment (anthonymccaigue) [inline packages/nist_registrar_server/src/handle_device_type_vulnerable.js:21]:
```suggestion
      return `device type ${deviceType} is already classified as vulnerable`;
```

File: packages/nist_registrar_server/src/handle_device_type_vulnerable.js [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    } else {
+      await dbRun("INSERT INTO has_vulnerability (device_type_id, vulnerability_id) VALUES (?, ?)",
+        [deviceTypeId, demoVulnerabilityId]);
+      return `device type ${deviceType} is now vulnerable`;
```
Comment (anthonymccaigue) [inline packages/nist_registrar_server/src/handle_device_type_vulnerable.js:25]:
```suggestion
      return `device type ${deviceType} is now classified as vulnerable`;
```

File: packages/nist_registrar_server/src/handle_device_type_vulnerable.js [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+    }
+  } else {
+    if (has_demo_vulnerability) {
+      await dbRun("DELETE FROM has_vulnerability WHERE (device_type_id = ? AND vulnerability_id = ?)",
+        [deviceTypeId, demoVulnerabilityId]);
+      return `device type ${deviceType} is now no longer vulnerable`;
+    } else {
+      return `device type ${deviceType} is already not vulnerable`;
```
Comment (anthonymccaigue) [inline packages/nist_registrar_server/src/handle_device_type_vulnerable.js:33]:
```suggestion
      return `device type ${deviceType} has already been classified not vulnerable`;
```

File: packages/nist_registrar_server/src/handle_manufacturer_trust.js [hunk 10]
```diff
@@ -0,0 +1,37 @@
+export default async function handle_manufacturer_trust(claimData, dbGet, dbRun) {
+  const {user, manufacturer, trust, issuanceDate} = claimData;
+  let manufacturerId = null;
+  console.log(`manufacturer: ${manufacturer}`)
```
Comment (anthonymccaigue) [inline packages/nist_registrar_server/src/handle_manufacturer_trust.js:4]: This log doesnt give much context

File: packages/nist_registrar_server/src/initialise_demo_database.js [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+      db.run("INSERT INTO gives_purchase_rights (recipient_id, authoriser_id, created_at) VALUES (?, ?, ?)", [ashsId, nicksIdForRights, new Date().toISOString()]);
+
+      db.run("INSERT INTO device_type (id, name, created_at) VALUES (?, ?, ?)", ['ac115a2d-4c18-4de4-9b59-4861ddb18b4e', 'Raspberry Pi', '2024-01-18 09:18:55.849129']);
+
+      db.run("INSERT INTO device_type (id, name, created_at) VALUES (?, ?, ?)", ['465df82c-d250-49c1-be27-95c8e4759fc2', 'Smart Speaker', '2024-01-16 14:13:55.849129']);
+
+      db.run("INSERT INTO vulnerability (id, name, severity, url, created_at) VALUES (?, ?, ?, ?, ?)", [demoVulnerabilityId, 'Security Flaw',	'Critical',	'https://example.com', '2024-01-16 14:13:55.849160']);
+      console.log("COMPLETE!!!!");
+    });
+    console.log("RETURNING DB");
```
Comment (anthonymccaigue) [inline packages/nist_registrar_server/src/initialise_demo_database.js:59]:
```suggestion
    });
```

File: packages/nist_registrar_server/src/server.js [hunk 12]
```diff
@@ -0,0 +1,263 @@
...
+      } catch (err) {
+        logs = [];
+      }
+      res.send(logs);
+    }
+    catch (err) {
+      console.log(`Encountered Error: ${err}`)
+      res.send(`Encountered Error: ${err}`);
+    }
+  }));
```
Comment (anthonymccaigue) [inline packages/nist_registrar_server/src/server.js:255]: Do you need to log and send the responses?

File: packages/nist_registrar_server/src/get_device_info.js [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+
+export default async function getDeviceInfo(device, dbGet) {
+  let deviceId = null;
+  let vulnerable = null;
+  let deviceType = null;
+  let trustingUser = null;
+  const deviceRow = await dbGet("SELECT id, name from device where id = ? OR name = ?", [device, device])
+  if (!deviceRow) {
+    console.log(`No device found for ID or name: ${device}`);
+    return `No device with id or name ${device}`;
```
Comment (anthonymccaigue) [inline packages/nist_registrar_server/src/get_device_info.js:11]:
```suggestion
    return `No device with id or name ${device}`;
```

File: packages/nist_registrar_server/src/get_device_type_info.js [hunk 14]
```diff
@@ -0,0 +1,15 @@
+const demoVulnerabilityId = '9dce9345-e306-4786-b7f0-536827351d21';
+
+export default async function getDeviceTypeInfo(deviceType, dbGet) {
+  let deviceTypeId = null;
+  const deviceTypeRow = await dbGet("SELECT id, name from device_type where id = ? OR name = ?", [deviceType, deviceType])
+  if (!deviceTypeRow) {
+    console.log(`No device type found for ID or name: ${deviceType}`);
+    return `No device type with id or name ${deviceType}`;
```
Comment (anthonymccaigue) [inline packages/nist_registrar_server/src/get_device_type_info.js:8]:
```suggestion
    return `No device type with id or name ${deviceType}`;
```

File: packages/nist_registrar_server/src/get_manufacturer_info.js [hunk 15]
```diff
@@ -0,0 +1,21 @@
+export default async function getManufacturerInfo(manufacturer, dbGet) {
+  let manufacturerId = null;
+  let trusted = false;
+  let trustingUser = null;
+  const manufacturerRow = await dbGet("SELECT id, name from manufacturer where id = ? OR name = ?", [manufacturer, manufacturer]);
+  if (!manufacturerRow) {
+    console.log(`No manufacturer found for ID or name: ${manufacturer}`);
+    return `No manufacturer with id or name ${manufacturer}`;
```
Comment (anthonymccaigue) [inline packages/nist_registrar_server/src/get_manufacturer_info.js:8]:
```suggestion
    return `No manufacturer with id or name ${manufacturer}`;
```

File: packages/nist_registrar_server/src/handle_device_trust.js [hunk 16]
```diff
@@ -0,0 +1,37 @@
+export default async function handleDeviceTrust(claimData, dbGet, dbRun) {
+  const {user, device, trust, issuanceDate} = claimData;
+  let deviceId = null;
+  console.log(`device: ${device}`)
+  const deviceRow = await dbGet("SELECT id from device where id = ? OR name = ?", [device, device])
+  if (!deviceRow) {
+    console.log(`No device found for ID or name: ${device}`);
+    return `No device with id or name ${device}`;
```
Comment (anthonymccaigue) [inline packages/nist_registrar_server/src/handle_device_trust.js:8]:
```suggestion
    return `No device with id or name ${device}`;
```

File: packages/nist_registrar_server/src/handle_device_trust.js [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+  } else {
+    deviceId = deviceRow.id;
+  }
+  let userId = null;
+  const userRow = await dbGet("SELECT id from user where id = ? OR username = ?", [user, user])
+  if (!userRow) {
+    console.log(`No user found for ID or name: ${user}`);
+    return `No user with id or name ${user}`;
```
Comment (anthonymccaigue) [inline packages/nist_registrar_server/src/handle_device_trust.js:16]:
```suggestion
    return `No user with id or name ${user}`;
```

File: packages/nist_registrar_server/src/handle_device_type_binding.js [hunk 18]
```diff
@@ -0,0 +1,28 @@
...
+    console.log(`No device found for ID or name: ${device}`);
+    return `No device with id or name ${device}`;
+  } else {
+    deviceId = deviceRow.id;
+  }
+  let deviceTypeId = null;
+  const deviceTypeRow = await dbGet("SELECT id from device_type where id = ? OR name = ?", [deviceType, deviceType])
+  if (!deviceTypeRow) {
+    console.log(`No device type found for ID or name: ${deviceType}`);
+    return `No device type with id or name ${deviceType}`;
```
Comment (anthonymccaigue) [inline packages/nist_registrar_server/src/handle_device_type_binding.js:16]:
```suggestion
    return `No device type with id or name ${deviceType}`;
```

File: packages/nist_registrar_server/src/handle_device_type_binding.js [hunk 19]
```diff
@@ -0,0 +1,28 @@
+export default async function handleDeviceTypeBinding(claimData, dbGet, dbRun) {
+  const {device, deviceType, issuanceDate} = claimData;device
+  let deviceId = null;
+  console.log(`device: ${device}`)
+  const deviceRow = await dbGet("SELECT id from device where id = ? OR name = ?", [device, device])
+  if (!deviceRow) {
+    console.log(`No device found for ID or name: ${device}`);
+    return `No device with id or name ${device}`;
```
Comment (anthonymccaigue) [inline packages/nist_registrar_server/src/handle_device_type_binding.js:8]:
```suggestion
    return `No device with id or name ${device}`;
```

File: packages/nist_registrar_server/src/handle_device_type_binding.js [hunk 20]
```diff
@@ -0,0 +1,28 @@
+export default async function handleDeviceTypeBinding(claimData, dbGet, dbRun) {
+  const {device, deviceType, issuanceDate} = claimData;device
+  let deviceId = null;
+  console.log(`device: ${device}`)
```
Comment (anthonymccaigue) [inline packages/nist_registrar_server/src/handle_device_type_binding.js:4]: This log doesn't give much context

File: packages/nist_registrar_server/src/handle_device_type_vulnerable.js [hunk 21]
```diff
@@ -0,0 +1,36 @@
+const demoVulnerabilityId = '9dce9345-e306-4786-b7f0-536827351d21';
+
+export default async function handleDeviceTypeVulnerable(claimData, dbGet, dbRun) {
+  const {deviceType, vulnerable, issuanceDate} = claimData;
+  let deviceTypeId = null;
+  const deviceTypeRow = await dbGet("SELECT id from device_type where id = ? OR name = ?", [deviceType, deviceType])
+  if (!deviceTypeRow) {
+    console.log(`No device type found for ID or name: ${deviceType}`);
+    return `No device type with id or name ${deviceType}`;
```
Comment (anthonymccaigue) [inline packages/nist_registrar_server/src/handle_device_type_vulnerable.js:9]:
```suggestion
    return `No device type with id or name ${deviceType}`;
```

File: packages/nist_registrar_server/src/handle_manufacturer_trust.js [hunk 22]
```diff
@@ -0,0 +1,37 @@
+export default async function handleManufacturerTrust(claimData, dbGet, dbRun) {
+  const {user, manufacturer, trust, issuanceDate} = claimData;
+  let manufacturerId = null;
+  console.log(`manufacturer: ${manufacturer}`)
+  const manufacturerRow = await dbGet("SELECT id from manufacturer where id = ? OR name = ?", [manufacturer, manufacturer]);
+  if (!manufacturerRow) {
+    console.log(`No manufacturer found for ID or name: ${manufacturer}`);
+    return `No manufacturer with id or name ${manufacturer}`;
```
Comment (anthonymccaigue) [inline packages/nist_registrar_server/src/handle_manufacturer_trust.js:8]:
```suggestion
    return `No manufacturer with id or name ${manufacturer}`;
```

File: packages/nist_registrar_server/src/handle_manufacturer_trust.js [hunk 23]
```diff
@@ -0,0 +1,37 @@
...
+  } else {
+    manufacturerId = manufacturerRow.id;
+  }
+  let userId = null;
+  const userRow = await dbGet("SELECT id from user where id = ? OR username = ?", [user, user])
+  if (!userRow) {
+    console.log(`No user found for ID or name: ${user}`);
+    return `No user with id or name ${user}`;
```
Comment (anthonymccaigue) [inline packages/nist_registrar_server/src/handle_manufacturer_trust.js:16]:
```suggestion
    return `No user with id or name ${user}`;
```

**Review by anthonymccaigue — COMMENTED** [review]
Hi Ash,
Code looks good but there are a lot of duplicated logs which may clog up the server

**Pull request merged**